### PR TITLE
LVUIDF-25: Image gallery

### PR DESCRIPTION
**Pull request by ladan12**
http://ssu-jira.softserveinc.com/browse/LVUIDF-25
Implement image gallery with uploader(carrierwave) and custom slider.
To review:
 - [x] @rtriska 
 - [ ] @forecome

To merge:
- [ ] @izagor 
- [ ] @forecome

### CODE REVIEW

File: app/assets/javascripts/components/albums/album.js.jsx [hunk 1]
```diff
@@ -10,63 +11,148 @@ var Album = React.createClass({
...
       }
-    });  
+    });
   },
   componentWillUnmount: function () {
     this.album.off('change');
   },
 
+  onDeleteAlbum: function(id) {
+    var self = this;
```
Comment (forecome) [inline app/assets/javascripts/components/albums/album.js.jsx]: Why do you need this variable? You can work with ```this```

File: app/assets/javascripts/components/albums/album.js.jsx [hunk 2]
```diff
@@ -10,63 +11,148 @@ var Album = React.createClass({
...
+    bootbox.confirm("Are you sure?", function(confirmation) {
+      if(confirmation == true) {
+        var album = new Bb.Models.Album({id: id}); {/*some starnge bugs here O_O. solved: write dataType: "text" */}
+        album.destroy({
+          success: function(model, response, options) {
+            self.transitionTo('albums');
+          }.bind(this),
+          error: function(model, response, options) {
+          }.bind(this),
```
Comment (forecome) [inline app/assets/javascripts/components/albums/album.js.jsx]: Do you need this empty callback?

File: app/assets/javascripts/components/albums/album.js.jsx [hunk 3]
```diff
@@ -10,63 +11,148 @@ var Album = React.createClass({
...
+        });
+      }
+      else {
+        return;
+      }
+    });
+  },
+
+  modalShow: function(picture) {
+    var self = this;
```
Comment (forecome) [inline app/assets/javascripts/components/albums/album.js.jsx]: The same question

File: app/assets/javascripts/components/albums/album.js.jsx [hunk 4]
```diff
@@ -10,63 +11,148 @@ var Album = React.createClass({
...
+              }
+            return false;
+          }
+        }
+      }
+    });
+  },
+
   render: function () {
+    var self = this;
```
Comment (forecome) [inline app/assets/javascripts/components/albums/album.js.jsx]: The same question. Also you are using ```this``` below. What the reason to create this variable?

File: app/assets/javascripts/components/albums/form.js.jsx [hunk 5]
```diff
@@ -0,0 +1,42 @@
+var AlbumForm = React.createClass({
+  handleFormSubmit: function(e) {
+    e.preventDefault();
+    var title = React.findDOMNode(this.refs.title).value.trim();
+    var description = React.findDOMNode(this.refs.description).value.trim();
+    var user_id = React.findDOMNode(this.refs.author).value.trim();
+    var album = {title: title, description: description, user_id: user_id};
```
Comment (forecome) [inline app/assets/javascripts/components/albums/form.js.jsx]: Please, use such syntax:
```
var a = 1,
    b = 2,
    c = 3;
```

File: app/assets/javascripts/components/albums/new.js.jsx [hunk 6]
```diff
@@ -0,0 +1,22 @@
+var AlbumNew = React.createClass({
+  mixins: [ReactRouter.Navigation],
+  handleSubmit: function(album) {
+  var alb = new Bb.Models.Album();
+    alb.save(album.album,{
+         success: function(data) {
+           this.transitionTo('/albums');
+         }.bind(this),
+         error: function(xhr, status, error) {
+         }.bind(this)
```
Comment (forecome) [inline app/assets/javascripts/components/albums/new.js.jsx]: Do you need this empty callback?

File: app/assets/javascripts/routes.js.jsx [hunk 7]
```diff
@@ -4,12 +4,23 @@ var DefaultRoute = ReactRouter.DefaultRoute,
...
     <Route path='/articles/:articleId/edit' name='edit_article' handler={EditArticle} />
+    <Route name='albumEdit' handler={AlbumEdit} path='/albums/:albumId/edit' />
+    {/*pictures
+      <Route name='pictures' handler={Pictures} path='/pictures' />
+    */}
+      <Route name='pictureNew' handler={PictureNew} path='/albums/:albumId/pictures/new' />
+    {/*
+      <Route name='picture' handler={Picture} path='/pictures/:pictureId' />
+      <Route name='pictureEdit' handler={PictureEdit} path='/pictures/:pictureId/edit' />
+    */}
```
Comment (forecome) [inline app/assets/javascripts/routes.js.jsx]: Please, remove commented code

File: app/uploaders/image_uploader.rb [hunk 8]
```diff
@@ -11,16 +11,18 @@ class ImageUploader < CarrierWave::Uploader::Base
   # Override the directory where uploaded files will be stored.
   # This is a sensible default for uploaders that are meant to be mounted:
   def store_dir
-    "uploads/#{model.class.to_s.underscore}/#{mounted_as}/#{model.id}"
+    # "uploads/#{model.class.to_s.underscore}/#{mounted_as}/#{model.id}"
+    'assets/images'
```
Comment (forecome) [inline app/uploaders/image_uploader.rb]: It is not correct. You should not place any uploaded files in assets folder. Please read about assets pipeline in Rails.
Also you remover unique parts of path (model name and model id). In this case if you upload two files with a same names you will get a conflict and file will be overwritten. Please, revert the changes.

File: app/uploaders/image_uploader.rb [hunk 9]
```diff
@@ -11,16 +11,18 @@ class ImageUploader < CarrierWave::Uploader::Base
...
   end
 
   # Provide a default URL as a default if there hasn't been a file uploaded:
-  # def default_url
+  def default_url
+    'assets/people.jpg'
```
Comment (forecome) [inline app/uploaders/image_uploader.rb]: You should use assets helpers. You even have example: see comments in lines 21-22.

File: app/assets/javascripts/components/albums/album.js.jsx (hunk 1 above)
Comment (ladan12) [inline app/assets/javascripts/components/albums/album.js.jsx]: I can't, because this and self are not equal in the scope of bootbox.confirm function

File: app/assets/javascripts/components/albums/album.js.jsx [hunk 10]
```diff
@@ -10,63 +11,148 @@ var Album = React.createClass({
...
-                   album = {this.state.album} /> 
-                   {this.state.pictures.map(function(picture) {
-                      return <div className='col-md-3' id='showAlbum'><img src='/uploads/picture/image/30/sample.jpg' key = {picture.id} /></div>;
-                   })}      
+                   album = {this.state.album}
+                   author = {this.state.album.user_id} />
+                   <div className='row'>
+                     {this.state.pictures.map(function(picture) {
+                        return <div className='col-md-3' id='showAlbum' key = {picture.id} ><img src={picture.image.url} onClick={self.modalShow.bind(this, picture)} /></div>;
+                     })}
```
Comment (ladan12) [inline app/assets/javascripts/components/albums/album.js.jsx]: I'm using self in line 115

File: app/assets/javascripts/components/albums/album.js.jsx (hunk 1 above)
Comment (forecome) [inline app/assets/javascripts/components/albums/album.js.jsx]: OK

File: app/assets/javascripts/components/albums/album.js.jsx (hunk 10 above)
Comment (forecome) [inline app/assets/javascripts/components/albums/album.js.jsx]: What is the difference between ```self``` and ```this``` in this particular method?

Comment (ladan12) [inline app/assets/javascripts/components/albums/album.js.jsx]: In this method ```this``` refers to Window object, whereas  ```self ```refers to ReactClassComponent object, which  contains  method modalShow. But I have a warning in my browser console, thus it's better to write like this:
```onClick={self.modalShow.bind(self, picture)}```
So now I can even write: 
```onClick={self.modalShow.bind(null, picture)}```

File: app/assets/javascripts/components/albums/album.js.jsx (hunk 1 above)
Comment (rtriska) [inline app/assets/javascripts/components/albums/album.js.jsx]: @ladan12 ask @Fleeck  about this issue. We solved it with '.bind(this)'

```javascript
bootbox.confirm("Are you sure?", function(confirmation) {
  if(confirmation == true) {
    var album = new Bb.Models.Album({id: id});
    album.destroy({
      success: function(model, response, options) {
        self.transitionTo('albums');
        console.log(self);
        console.log(this);
      }.bind(this),
      dataType: "text"
    });
  }
  else {
    return;
  }
}.bind(this));
```

File: spec/controllers/api/v1/albums_controller_spec.rb [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+    context 'valid attributes' do
+      it 'should update album' do
+        patch :update, {id: @album.id, album: {title: "new title", description: "new description"}}, format: :json
+        @album.reload
+        expect(response).to have_http_status(:ok)
+        expect(@album.title).to eq('new title')
+        expect(@album.description).to eq('new description')
+
+        content = response.body
+        expect(content).to include 'title'
```
Comment (rtriska) [inline spec/controllers/api/v1/albums_controller_spec.rb]: For future: 1 expectation per 1 test

**Comment by rtriska** [comment]
@ladan12 Why do you ignore all comments? 
Why do you commit Gemfile.lock? Never do this!
Why do you commit images? You have AWS and can upload any of needed images!

@khomyakov could you please check and update this PR according to comments? Also please check failing tests. Ensure that everything works properly.

File: app/assets/javascripts/components/pictures/form.js.jsx [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+    newState[e.target.id] = e.target.value;
+    this.setState(newState);
+  },
+  handleSubmit: function(e) {
+    e.preventDefault();
+    var form = document.forms.namedItem("myForm"),
+        data = new FormData(form),
+        oReq = new XMLHttpRequest();
+    oReq.open("POST", "api/v1/pictures", true);
+    oReq.send(data);
```
Comment (yurakuzh) [inline app/assets/javascripts/components/pictures/form.js.jsx]: why don't 
$.ajax  ?

File: app/assets/javascripts/components/albums/album.js.jsx [hunk 13]
```diff
@@ -10,63 +11,126 @@ var Album = React.createClass({
...
+      }
+    }.bind(this));
+  },
+
+  modalShow: function(picture) {
+    var self = this;
+    var pictures = self.state.album.pictures;
+    var index;
+            self.state.album.pictures.forEach(function(pic){
+              if(pic.image.url == picture.image.url) {
```
Comment (yurakuzh) [inline app/assets/javascripts/components/albums/album.js.jsx]: use ===

File: app/assets/javascripts/components/albums/edit.js.jsx [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+    };
+  },
+  componentDidMount: function () {
+    this.album = new Bb.Models.Album({id: this.props.params.albumId});
+    this.album.fetch({
+      success: function(model, response, options) {
+        this.setState({ album: model.toJSON(), pictures: model.toJSON().pictures});
+      }.bind(this),
+      error: function(model, response, options) {
+        console.log(response.status);
```
Comment (yurko) [inline app/assets/javascripts/components/albums/edit.js.jsx]: please remove ```console.log()```

File: app/assets/javascripts/components/pictures/edit.js.jsx [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+    var pic = new Bb.Models.Picture({id: this.props.params.pictureId});
+    pic.save(picture, {patch: true,
+      success: function(model, response, options) {
+        this.setState({ picture: model.toJSON()});
+        this.transitionTo('/albums/' + this.props.params.albumId + '/pictures');
+      }.bind(this)
+    });
+  },
+  render: function() {
+      if(this.state.picture.image) {console.log(this.state.picture.image.url)};
```
Comment (yurko) [inline app/assets/javascripts/components/pictures/edit.js.jsx]: do we need ```console.log()``` here?

File: app/assets/javascripts/components/pictures/form.js.jsx [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+    //   }.bind(this)
+    // });
+    this.setState({picture: pic});
+  },
+  componentWillReceiveProps: function(nextProps) {
+    this.setState({picture: nextProps.picture});
+  },
+  handleChange: function(e) {
+    var newState = this.state.picture;
+    console.log(newState);
```
Comment (yurko) [inline app/assets/javascripts/components/pictures/form.js.jsx]: please remove ```console.log()```
